frappe/api.py

```diff
@@ -167,23 +167,38 @@ def validate_auth_via_api_keys():
 	"""
 	try:
 		authorization_header = frappe.get_request_header("Authorization", None).split(" ") if frappe.get_request_header("Authorization") else None
+		authorization_source = frappe.get_request_header("Frappe-Authorization-Source", None) if frappe.get_request_header("Frappe-Authorization-Source") else None
 		if authorization_header and authorization_header[0] == 'Basic':
 			token = frappe.safe_decode(base64.b64decode(authorization_header[1])).split(":")
-			validate_api_key_secret(token[0], token[1])
+			validate_api_key_secret(token[0], token[1], authorization_source)
 		elif authorization_header and authorization_header[0] == 'token':
 			token = authorization_header[1].split(":")
-			validate_api_key_secret(token[0], token[1])
+			validate_api_key_secret(token[0], token[1], authorization_source)
 	except Exception as e:
 		raise e
 
-def validate_api_key_secret(api_key, api_secret):
-	user = frappe.db.get_value(
-		doctype="User",
-		filters={"api_key": api_key},
-		fieldname=['name']
-	)
-	form_dict = frappe.local.form_dict
-	user_secret = frappe.utils.password.get_decrypted_password ("User", user, fieldname='api_secret')
-	if api_secret == user_secret:
-		frappe.set_user(user)
-		frappe.local.form_dict = form_dict
+def validate_api_key_secret(api_key, api_secret, frappe_authorization_source=None):
+	""" frappe_authorization_source to provide api key and secret for a doctype apart from User """
+	if frappe_authorization_source:
```

Comment:
This will get rid of the `if` block

```
doctype = frappe_authorization_source or 'User'
```

Also, since this is an alternate method of authentication, should be documented.

Reply:
https://github.com/frappe/frappe_io/pull/237